### PR TITLE
NestedFolders: Virtual tree view

### LINKED ISSUE #65683: NestedFolders: Browse dashboards nested tree component

### PR DESCRIPTION
**Pull request by joshhunt**
Early pass at the new Browse Dashboards view.

To simplify state management (so we don't have to actually keep around a tree structure that's annoying to update), data is stored in a relatively flat structure.

 - BrowseView fetches data and owns all state.
   - Keeps lists of children in an object keyed by folderUID. 
   - Folder open state is kept in a seperate object (keyed by folderUID).
   - As we iterate, we might want to combine these into one object, but im not convinced we need that yet
 - DashboardsTree accepts a specially formatted list and displays the folder + dashboard heirachy in a tree-like list/table.
   - Accepts a 'tree' as a flat array, but where each item has a 'level' so it can render with the right indentation.

image

What should work:
 - Browse root dashboards + folders
 - Browse folder view
 - Expand folders to show nested items (for whole hierarchy)
 - Clicking folder or dashboard navigates to that item

Item selection is not implemented. Will probably happen in seperate PR.

Closes https://github.com/grafana/grafana/issues/65683

### CODE REVIEW

File: public/app/features/browse-dashboards/components/DashboardsTree.tsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        return (
+          <div key={key} {...headerGroupProps} className={styles.headerRow}>
+            {headerGroup.headers.map((column) => {
+              const { key, ...headerProps } = column.getHeaderProps();
+
+              return (
+                <div key={key} {...headerProps} role="columnheader" className={styles.cell}>
+                  {column.render('Header')}
+                </div>
+              );
```
Comment (tskarhed) [inline public/app/features/browse-dashboards/components/DashboardsTree.tsx:70]: Maybe these can be extracted to separate functions for better readability?

Comment (joshhunt) [reply public/app/features/browse-dashboards/components/DashboardsTree.tsx]: I've been thinking the same thing, so I gave it a try but I don't think it aided much. I think one of the main problems is how eslint needs us to explicitly set key on the map children - it cannot see that getHeaderProps returns key in its type, so we still need to do the same destructure and spread, which I think is the main part that makes this tricky to read.

Comment (joshhunt) [reply public/app/features/browse-dashboards/components/DashboardsTree.tsx]: If people feel strongly about this happy to take another pass.

File: public/app/features/browse-dashboards/components/TypeCell.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          <Icon name={iconName} /> Panel
+        </Text>
+      );
+    default:
+      return null;
+  }
+}
+
+// TODO: Grafana UI????
+interface TextProps {
```
Comment (joshhunt) [inline public/app/features/browse-dashboards/components/TypeCell.tsx:39]: I will remove this in favor of just using the style directly

File: public/app/features/browse-dashboards/components/DashboardsTree.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+const ROW_HEIGHT = 35;
+
+export function DashboardsTree({ items, width, height, onFolderClick }: DashboardsTreeProps) {
+  const styles = useStyles2(getStyles);
+
+  const tableColumns = useMemo(() => {
+    const checkboxColumn: DashboardsTreeColumn = {
+      id: 'checkbox',
+      Header: () => <Checkbox value={false} />,
+      Cell: () => <Checkbox value={false} />,
```
Comment (tskarhed) [inline public/app/features/browse-dashboards/components/DashboardsTree.tsx:33]: Currently the checkboxes don't do anything. Should a handler be passed here?

Comment (joshhunt) [reply public/app/features/browse-dashboards/components/DashboardsTree.tsx:33]: > Item selection is not implemented. Will probably happen in seperate PR.

**Review by tskarhed — APPROVED** [review]
Looks good to me, maybe someone else can have a look too

**Review by ashharrison90 — APPROVED** [review]
lgtm 🙌

**Pull request merged**